reducerfunction.py

```diff
@@ -0,0 +1,163 @@
+import copy
+import json
+import logging
+import multiprocessing
+import os
+import random
+import shutil
+import time
+from pathlib import Path
+from typing import Any, Optional
+
+import ccbuilder
+import diopter
+
+import preprocessing
+import utils
+
+
+def reduce_file(
+    config: utils.NestedNamespace,
+    bldr: ccbuilder.Builder,
+    file: Path,
+    force: bool = False,
+    jobs: int = multiprocessing.cpu_count(),
+) -> bool:
+    """Reduce a case given in the .tar format.
+    Interface for `reduced_code`.
+
+    Args:
+        file (Path): Path to .tar case.
+        force (bool): Force a reduction (even if the case is already reduced).
+    Returns:
+        bool: If the reduction was successful.
+    """
+    case = utils.Case.from_file(config, file)
+
+    if reduce_case(config=config, bldr=bldr, case=case, force=force, jobs=jobs):
+        case.to_file(file)
+        return True
+    return False
+
+
+def reduce_case(
+    config: utils.NestedNamespace,
+    bldr: ccbuilder.Builder,
+    case: utils.Case,
+    force: bool = False,
+    jobs: int = multiprocessing.cpu_count(),
+) -> bool:
+    """Reduce a case.
+
+    Args:
+        case (utils.Case): Case to reduce.
+        force (bool): Force a reduction (even if the case is already reduced).
+
+    Returns:
+        bool: If the reduction was successful.
+    """
+    if not force and case.reduced_code:
+
+        return True
+
+    case.reduced_code = reduce_code(
+        config=config,
+        bldr=bldr,
+        code=case.code,
+        marker=case.marker,
+        bad_setting=case.bad_setting,
+        good_settings=case.good_settings,
+        bisection=case.bisection,
+        jobs=jobs,
+    )
+    return bool(case.reduced_code)
+
+
+def reduce_code(
```

Comment:
This can be simplified a lot with the following changes:

- Let the caller handle preprocessing
- Instead of creating a script to call the checker, we expose the relevant functionality there and we use `diopter.make_interestingness_test`. This will also get rid of the temprorary files